File: QA_Interview_Questions/03-reverse-string.ts
/**
 * Question 3: Reverse String
 * Problem: Write a function to reverse a string in-place (without using built-in reverse)
 * 
 * Example:
 * Input: "hello"
 * Output: "olleh"
 */

function reverseString(s: string): string {
  let result = '';
  for (let i = s.length - 1; i >= 0; i--) {
    result += s[i];
  }
  return result;
}

// Alternative: Using spread operator
function reverseStringAlternative(s: string): string {
  return [...s].reverse().join('');
}

// Alternative: Using recursion
function reverseStringRecursive(s: string): string {
  if (s.length === 0) return '';
  return reverseStringRecursive(s.slice(1)) + s[0];
}

// Test cases
console.log(reverseString("hello")); // "olleh"
console.log(reverseString("TypeScript")); // "tpircSepyT"
console.log(reverseString("123")); // "321"
console.log(reverseStringAlternative("world")); // "dlrow"
console.log(reverseStringRecursive("abc")); // "cba"

File: Java_Interview_Questions/03-ReverseString.java
/**
 * Question 3: Reverse String
 * Problem: Reverse a string without using built-in reverse methods
 * 
 * Example:
 * Input: "hello"
 * Output: "olleh"
 */

public class ReverseString {
    
    public static String reverseString(String str) {
        String result = "";
        for (int i = str.length() - 1; i >= 0; i--) {
            result += str.charAt(i);
        }
        return result;
    }
    
    // Using StringBuilder (more efficient)
    public static String reverseStringBuilder(String str) {
        return new StringBuilder(str).reverse().toString();
    }
    
    // Using character array
    public static String reverseStringArray(String str) {
        char[] chars = str.toCharArray();
        int left = 0, right = chars.length - 1;
        
        while (left < right) {
            char temp = chars[left];
            chars[left] = chars[right];
            chars[right] = temp;
            left++;
            right--;
        }
        return new String(chars);
    }
    
    // Using recursion
    public static String reverseStringRecursive(String str) {
        if (str.isEmpty()) {
            return str;
        }
        return reverseStringRecursive(str.substring(1)) + str.charAt(0);
    }
    
    public static void main(String[] args) {
        System.out.println(reverseString("hello")); // "olleh"
        System.out.println(reverseStringBuilder("TypeScript")); // "tpircSepyT"
        System.out.println(reverseStringArray("123")); // "321"
        System.out.println(reverseStringRecursive("abc")); // "cba"
    }
}
